Custom attributes

MosheWolberg opened the pull request:
This is the core functionality for Custom Attributes.

At present, only Custom Attributes on types are supported but that extending to Assemblies, Members etc. shouldn't be difficult since the foundation is in place.
I reorganized several key parts of the projects:

1. Most storage and all construction of metadata was centralized into the ModuleBuilder class.
This made more sense than the previous local construction the metadata of each type, member etc within itself - a lot of metadata across types may overlap (for example, assembly refs) so constructing data in the central module made sense and also simplifies the code going forward.

2.  Actual metadata writing moved to static Metadata Helper class.

3. Signature construction moved to Signature Helper class. 

And then I added necessary code for Custom Attributes, (porting needed encoder classes as well like Custom Attribute Builder).

Fields are next but I want to tie up a few loose ends before then (I'll share a list of them) - mainly, the SignatureEncoder class is currently just a basic prototype I built much earlier but to support more complex things like arrays in signatures I should port the real SignatureEnconder.

MosheWolberg committed:
Writing an empty assembly to disk and reading it back

MosheWolberg committed:
A basic assembly write and read

MosheWolberg committed:
Style fixes from comments

MosheWolberg committed:
Minor changes + fixed metadata to include Version etc.

MosheWolberg committed:
Further style fixes

MosheWolberg committed:
Style Chnage

MosheWolberg committed:
Other files

MosheWolberg committed:
Fix Build failure

MosheWolberg committed:
I am still learning Git and switched from my new functionality branch into this one without commiting the new changes and it caused those changes to be added into my old branch. So I am just going to push everything so that the build works. Take a look if you want, its just blank interfaces being added to types but methods in the interfaces are still to come.

MosheWolberg committed:
Support for interfaces and their members.

MosheWolberg committed:
Reflection Testing, Removed LinkedLists, Small Changes

MosheWolberg committed:
Rebasing

MosheWolberg committed:
Move MetadataBuilder var and typo fixes

MosheWolberg committed:
Sharing work

MosheWolberg committed:
Work in progress on Custom Attributes

MosheWolberg committed:
More progress

MosheWolberg committed:
Milestone Commit - Custom Attributes

MosheWolberg committed:
Removed reference to local dll breaking build

MosheWolberg commented:
@joperezr

MosheWolberg commented:
Hey, not urgent, not to ping or pressure anyone, I believe it's a very busy time with preview 7, - but just tagging you guys because it doesn't look like you were automatically tagged.
@ericstj @jkotas @AaronRobinsonMSFT

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+{
+    public class CustomAttributeBuilder
+    {
+        internal ConstructorInfo m_con;
```
NIT: Why do we have `m_` prefixes for these? I'd assume these should all just be prefixed by `_`. Also, for this particular one, can we name it `_constructorInfo`?

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+    {
+        internal ConstructorInfo m_con;
+        private object?[] m_constructorArgs;
+        internal byte[] m_blob;
```
NIT: it would be good if we can also add some comments that explain what is this used for, and why it's internal instead of private.

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+        private object?[] m_constructorArgs;
+        internal byte[] m_blob;
+
+        public ConstructorInfo Con { get => m_con;  }
```
`Con` => `Constructor`. Also, where is it used?

joperezr replied:
Same goes for all of the places where you take a parameter named `con`, we should just instead use `constructorInfo`

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+        public ConstructorInfo Con { get => m_con;  }
+
+        // public constructor to form the custom attribute with constructor and constructor
+        // parameters.
```
Instead of adding  normal comments `//`, you could just add tripple-slash comments instead `///` which VS will automatically template for you so that you get a nice intellisense experience when consuming these.

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+            ArgumentNullException.ThrowIfNull(fieldValues);
+
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly, combination of arguments used
+            if (namedProperties.Length != propertyValues.Length)
+                throw new ArgumentException("namedProperties, propertyValues differ");
+            if (namedFields.Length != fieldValues.Length)
+                throw new ArgumentException("namedFields, fieldValues differ");
```
```suggestion
            if (namedProperties.Length != propertyValues.Length)
            {
                throw new ArgumentException($"{nameof(namedProperties)} and {nameof(propertyValues)} should have the same length.");
            }

            if (namedFields.Length != fieldValues.Length)
            {
                throw new ArgumentException($"{nameof(namedFields)} and {nameof(fieldValues)} should have the same length.");
            }
```

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+
+            if ((con.Attributes & MethodAttributes.Static) == MethodAttributes.Static ||
+                (con.Attributes & MethodAttributes.MemberAccessMask) == MethodAttributes.Private)
+                throw new ArgumentException("Bad Constructor");
```
Can we provide better exception messages like: `The passed-in constructor is either static or private`

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+                    case TypeCode.Int64:
+                    case TypeCode.UInt64:
+                        return true;
+                    default:
```
Is this just the case for nint and nuint?

MosheWolberg replied:
This is straight from Reflection.Emit.

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+            // Verify that the constructor has a valid signature (custom attributes only support a subset of our type system).
+            for (i = 0; i < paramTypes.Length; i++)
+                if (!ValidateType(paramTypes[i]))
+                    throw new ArgumentException("Bad Type In Custom Attribute");
```
Should we mention here which one was the one at fault?

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+            }
+
+            // Allocate a memory stream to represent the CA blob in the metadata and a binary writer to help format it.
+            MemoryStream stream = new MemoryStream();
```
shouldn't these two be wrapped in a using statement to get them disposed?

MosheWolberg replied:
This is straight from Reflection.Emit.

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+                if (!property.CanWrite)
+                    throw new ArgumentException("Not A Writable Property");
+
+                //// Property has to be from the same class or base class as ConstructorInfo. Note: This edge case is currently difficult to port and needs to be added into prototype.
```
Should we remove this for now and instead have an issue tracking this?

joperezr commented on src/Experiment/src/CustomAttributeBuilder.cs:
```diff
@@ -0,0 +1,581 @@
+                if (!ValidateType(fldType))
+                    throw new ArgumentException("Bad Type In Custom Attribute");
+
+                // Field has to be from the same class or base class as ConstructorInfo. Note: This edge case is currently difficult to port and needs to be added into prototype.
```
Same here, perhaps remove this and just track it in an issue.

AaronRobinsonMSFT commented on src/Experiment/src/SignatureHelper.cs:
```diff
@@ -0,0 +1,94 @@
+    //For more complex signatures, port System.Reflection.Emit's SignatureHelper.
+    internal static class SignatureHelper
+    {
+        internal static BlobBuilder MethodSignatureEnconder(ParameterInfo[]? parameters, ParameterInfo? returnType, bool isInstance)
```
```suggestion
        internal static BlobBuilder MethodSignatureEncoder(ParameterInfo[]? parameters, ParameterInfo? returnType, bool isInstance)
```

Several places have the same typo.
```suggestion
        internal static BlobBuilder MethodSignatureEnconder(ParameterInfo[]? parameters, ParameterInfo? returnType, bool isInstance)
```

AaronRobinsonMSFT commented on src/Experiment/src/SignatureHelper.cs:
```diff
@@ -0,0 +1,94 @@
+            ParametersEncoder _parEncoder;
+            ReturnTypeEncoder _retEncoder;
+
+            new BlobEncoder(methodSignature).
```
I'd place the `.` on the next line.

AaronRobinsonMSFT commented on src/Experiment/src/TypeNameBuilder.cs:
```diff
@@ -0,0 +1,346 @@
+            _instNesting++;
+            _firstInstArg = true;
+
+            if (_useAngleBracketsForGenerics)
+                Append('<');
+            else
+                Append('[');
```
Seems common enough to make a single method.

AaronRobinsonMSFT replied:
Could be combined to do both open and close.

AaronRobinsonMSFT commented on src/Experiment/tests/ReflectionTesting.cs:
```diff
@@ -0,0 +1,465 @@
+        const bool _keepFiles = true; // keep files after testing for inspection
+        TempFileCollection _tfc;
+
+        [Fact]
```
I mentioned this previously. This is not the correct way to setup a test class asset. See https://xunit.net/docs/shared-context

MosheWolberg commented:
I've addressed these issues in my code. Since mostly small, added them into new work I've been doing, instead of writing them twice, to this branch and new branch (A lot of the changes in files overlap between branches). Will push them as part them as part next feature PR. 
Can we merge?

joperezr commented:
I think if possible it would be better to add the fixes here, that way there is no un-relevant code changes in your next PR. I'm happy to work with you offline to get these changes here.

MosheWolberg committed:
Feedback

MosheWolberg committed:
WIP

MosheWolberg committed:
Adressed feedback

MosheWolberg commented:
Merge?
